text/0000-hash/0000-hash.md

```diff
@@ -0,0 +1,302 @@
+let transaction = [0i8; 8019];
+let mut tx_hash = [0i8; 243];
+
+// Absorb the transaction.
+curl.absorb(&transaction);
+// And squeeze into the `tx_hash` buffer.
+curl.squeeze_into(&mut tx_hash);
+```
+
+```rust
+// Create a default Kerl instance. Kerl does not come with a configurable number of rounds, so that
+// `Kerl::new()` and `Kerl::default()` yield the same result.
+let mut kerl = Kerl::default();
+
+// `Kerl::digest` is a function that combines `Kerl::absorb` and `Kerl::squeeze`. `Kerl::digest_into`
+// combines `Kerl::absorb` with `Kerl::squeeze_into`. `CurlP` provides the same methods.
+let tx_hash = kerl.digest(&transaction);
+```
+
+Useful links:
+
+- [IOTA transactions and bundle](https://docs.iota.org/docs/dev-essentials/0.1/concepts/bundles-and-transactions)
+
+# Detailed design
+
+The sponges proposes in this RFC have a common interface, which consists of the functions below:
+
++ `fn absorb(&mut self, input: &[Trit])`;
++ `fn squeeze(&mut self) -> Vec<Trit>`;
++ `fn squeeze_into(&mut self, output: &mut [Trit])`;
++ `fn reset(&mut self)`;
++ `fn digest(&mut self, input: &[Trit]) -> Vec<Trit>`;
++ `fn digest_into(&mut self, input: &[Trit], output: &mut [Trit])`.
+
+This design is going to be the common interface of any sponge function. Users can expect each hash instance to provide
+at least these public methods. Following the sponge metaphor, an input reference provided by the user is `absorb`ed, and
+an output will be `squeeze`d from the data structure. `digest` is a convenience method calling `absorb` and `squeeze` in
+one go. The `*_into` versions of these methods are for providing a buffer into which the calculated hashes are written.
+The internal state will not be cleared unless `reset` is called. 
+
+## Detailed design using the `CurlP` example
+
+Using `CurlP` as an example, this section demonstrates what an implementation would look like.
+
+```rust
+#[derive(Clone, Copy)]
+pub struct CurlP {
+    /// The number of rounds of hashing to apply before a hash is squeezed.
+    rounds: usize,
+
+    /// The internal state.
+    state: [Trit; STATE_LENGTH],
+}
+```
+
+The default implementation of `CurlP` is always assumed to use 81 rounds:
+
+```rust
+impl Default for CurlP {
+    fn default() -> Self {
+        Self::new(81)
+    }
+}
+```
+
+```rust
+impl CurlP {
+
+    /// Absorb trits into the sponge
+    pub fn absorb(&mut self, input: &[Trit]) {
+        for c in input.chunks(HASH_LENGTH) {
+            self.state[0..c.len()].copy_from_slice(c);
+            self.transform();
+        }
+    }
+
+    pub fn digest(&mut self, input: &[Trit]) -> Vec<Trit> {
+        self.absorb(input);
+        self.squeeze()
+    }
+
+    /// Digest inputs and then compute the hash with length of provided output slice
+    pub fn digest_into(&mut self, input: &[Trit], buf: &mut [Trit]) {
+        self.absorb(input);
+        self.squeeze_into(output);
+    }
+
+    pub fn new(rounds: usize) -> Self {
```

Comment:
Should these be discrete values depending on a enum ?

Reply:
In the discussion on discord we mentioned const generics. But consider it will still be unstable for another year. We stay with this and wrap it with newtypes.